server/PathServices/SignalRRealtimeDataRepository.cs

```diff
@@ -0,0 +1,236 @@
+﻿namespace PathApi.Server.PathServices
+{
+    using Microsoft.AspNetCore.SignalR.Client;
+    using Newtonsoft.Json;
+    using PathApi.Server.PathServices.Models;
+    using PathApi.V1;
+    using Serilog;
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// A self-updating repository containing the realtime predicted arrivals of PATH trains sourced from SignalR hubs.
+    /// </summary>
+    internal sealed class SignalRRealtimeDataRepository : IRealtimeDataRepository, IDisposable
+    {
+        private readonly IPathDataRepository pathDataRepository;
+        private readonly IPathApiClient pathApiClient;
+        private readonly ConcurrentDictionary<(Station, RouteDirection), HubConnection> hubConnections;
+        private readonly ConcurrentDictionary<(Station, RouteDirection), List<RealtimeData>> realtimeData;
+
+        /// <summary>
+        /// Constructs a new instance of the <see cref="SignalRRealtimeDataRepository"/>.
+        /// </summary>
+        public SignalRRealtimeDataRepository(IPathDataRepository pathDataRepository, IPathApiClient pathApiClient)
+        {
+            this.pathDataRepository = pathDataRepository;
+            this.pathApiClient = pathApiClient;
+            this.hubConnections = new ConcurrentDictionary<(Station, RouteDirection), HubConnection>();
+            this.realtimeData = new ConcurrentDictionary<(Station, RouteDirection), List<RealtimeData>>();
+
+            this.pathDataRepository.OnDataUpdate += this.PathSqlDbUpdated;
+        }
+
+        /// <summary>
+        /// Gets the latest realtime train arrival data for the specified station.
+        /// </summary>
+        /// <param name="station">The station to get realtime arrival data for.</param>
+        /// <returns>A collection of arriving trains.</returns>
+        public Task<IEnumerable<RealtimeData>> GetRealtimeData(Station station)
+        {
+            return Task.FromResult(this.GetRealtimeData(station, RouteDirection.ToNY).Union(this.GetRealtimeData(station, RouteDirection.ToNJ)).Where(data => data.DataExpiration > DateTime.UtcNow));
```

Comment:
I think you're right that Concat is better than Union here as we don't need it to check for duplicates and deal with deduping. (IIUC Union will remove duplicates)